components/shift_reg/shift_reg.c

```diff
@@ -0,0 +1,203 @@
+
+    switch (dev->mode.dir)
+    {
+        case SHIFT_DIR_OUTPUT:
+            // set as output mode
+            io_conf.mode = GPIO_MODE_OUTPUT;
+            uint32_t buf32_0 = 0;
+            uint32_t buf32_1 = 0;
+            uint64_t result = 0;
+
+            if (dev->pin.clk >= 32)
+                buf32_1 |= 1 << (dev->pin.clk - 32);
+            else
+                buf32_0 |= 1 << dev->pin.clk;
+
+            if (dev->pin.latch >= 32)
+                buf32_1 |= 1 << (dev->pin.latch - 32);
+            else
+                buf32_0 |= 1 << dev->pin.latch;
+
+            if (dev->pin.data >= 32)
+                buf32_1 |= 1 << (dev->pin.data - 32);
+            else
+                buf32_0 |= 1 << dev->pin.data;
+
+            result = ((uint64_t)buf32_1 << 32) | ((uint64_t)buf32_0 << 0);
+
+            io_conf.pin_bit_mask = result;
+
+            break;
+
+        default:
+            ESP_LOGE(tag, "%s: Mode of shift register not found", __func__);
+            err = ESP_ERR_INVALID_ARG;
+            break;
+    }
+    // disable pull-down mode
+    io_conf.pull_down_en = 0;
+    // disable pull-up mode
+    io_conf.pull_up_en = 0;
+    // configure GPIO with the given settings
+    err = gpio_config(&io_conf);
+
+    return err;
+}
+
+esp_err_t shift_reg_deinit(shift_reg_config_t *dev)
+{
+    if (dev == NULL)
+    {
+        ESP_LOGE(tag, "%s: must have a valid argument;", __func__);
+        return ESP_ERR_INVALID_ARG;
+    }
+
+    free(dev->reg_value);
+    return ESP_OK;
+}
+
+esp_err_t shift_reg_send(shift_reg_config_t *dev, uint8_t *data, uint8_t len)
+{
+    esp_err_t err = ESP_FAIL;
+
+    if (dev == NULL || len > dev->num_reg || data == NULL)
+    {
+        ESP_LOGE(tag, "%s: must have a valid argument;", __func__);
+        err = ESP_ERR_INVALID_ARG;
+        return err;
+    }
+
+    if (dev->mode.bit_mode == SHIFT_BIT_MODE_MSB)
+    {
+        for (uint8_t i = 0; i < len; i++)
+        {
+            shift_reg_send8bits(dev, data[i]);
+            dev->reg_value[i] = data[i];
+        }
+    }
+    else
+    {
+        for (int8_t i = len - 1; i >= 0; i--)
+        {
+            shift_reg_send8bits(dev, data[i]);
+            dev->reg_value[i] = data[i];
+        }
+    }
+
+    err = ESP_OK;
+
+    return err;
+}
+
+esp_err_t shift_reg_send8bits(shift_reg_config_t *dev, uint8_t data)
+{
+    esp_err_t err = ESP_FAIL;
+
+    if (dev == NULL)
+    {
+        ESP_LOGE(tag, "%s: must have a valid argument;", __func__);
+        err = ESP_ERR_INVALID_ARG;
+        return err;
+    }
+
+    if (dev->mode.bit_mode == SHIFT_BIT_MODE_MSB)
+    {
+        // MSB Mode
+        for (int8_t i = 7; i >= 0; i--)
+        {
+            if ((data >> i) & 1)
+            {
+                gpio_set_level(dev->pin.data, true);
+            }
+            else
+            {
+                gpio_set_level(dev->pin.data, false);
+            }
+
+            gpio_set_level(dev->pin.clk, true);
```

Comment:
Perhaps it would be better to call the function `shift_reg_latch()` here

Reply:
Hi @UncleRus, sorry for just now fixing those open points.

I'm not sure what you had asked here.... so do you want me to latch right after a single bit has being transfered to the shift register?
Something like this:
```c
// MSB Mode
for (int8_t i = 7; i >= 0; i--)
{
    if ((data >> i) & 1)
    {
        gpio_set_level(dev->pin.data, true);
    }
    else
    {
        gpio_set_level(dev->pin.data, false);
    }

    gpio_set_level(dev->pin.clk, true);
    ets_delay_us(1);
    gpio_set_level(dev->pin.clk, false);
    ets_delay_us(1);
    shift_reg_latch(dev);
}
```

If so, I don't think that would make sense, since some times we need to more than a single bit at once. So let the user choose when perform the latch would be the best choice, in my point of view.

Please, let me know what you think or clarify what was your request on that, then I will fix that in a future commit.